server/src/main/java/org/elasticsearch/node/NodeConstruction.java

```diff
@@ -990,6 +986,9 @@ public Map<String, String> queryFields() {
             .map(TerminationHandlerProvider::handler);
         terminationHandler = getSinglePlugin(terminationHandlers, TerminationHandler.class).orElse(null);
 
+        if (clusterInfoService instanceof InternalClusterInfoService icis) {
+            icis.setHeapUsageSupplier(serviceProvider.newHeapUsageSupplier(pluginsService));
+        }
```

Comment:
I really don't love this, I will dig into whether there's a nicer/more idiomatic way to do it.

Reply:
What I had in mind is to set this in `Stateless#createComponents`, something like
```java
services.allocationService().setHeapUsageSupplier();
```

`AllocationService` has a reference to `ClusterInfoService` and can then handle it internally. This avoids changes to the Plugin interface which I think it is better since we might want to move the heap decider to stateful in future.

Reply:
Done in 23eb8e61036f7ef0b9e6246c9b975d3dded01ed6